wen-restart/src/wen_restart.rs

```diff
@@ -1,59 +1,307 @@
+    },
     solana_ledger::{ancestor_iterator::AncestorIterator, blockstore::Blockstore},
+    solana_program::{clock::Slot, hash::Hash},
+    solana_runtime::bank_forks::BankForks,
+    solana_sdk::timing::timestamp,
     solana_vote_program::vote_state::VoteTransaction,
     std::{
-        fs::File,
-        io::{Error, Write},
+        collections::{HashMap, HashSet},
+        fs::{read, File},
+        io::{Cursor, Write},
         path::PathBuf,
-        sync::Arc,
+        str::FromStr,
+        sync::{
+            atomic::{AtomicBool, Ordering},
+            Arc, RwLock,
+        },
+        thread::sleep,
+        time::Duration,
     },
+    thiserror::Error,
 };
 
-pub fn wait_for_wen_restart(
-    wen_restart_path: &PathBuf,
+// If >42% of the validators have this block, repair this block locally.
+const REPAIR_THRESHOLD: f64 = 0.42;
+
+#[derive(Error, Debug)]
+pub enum WenRestartError {
+    #[error("Protobuf Decode error: {0}")]
+    DecodeError(#[from] prost::DecodeError),
+    #[error("Protobuf Encode error: {0}")]
+    EncodeError(#[from] prost::EncodeError),
+    #[error("Exiting")]
+    Exiting,
+    #[error("Invalid Last Vote Type: {0:?}")]
+    InvalidLastVoteType(VoteTransaction),
+    #[error("File IO error: {0}")]
+    IoError(#[from] std::io::Error),
+    #[error("Missing last voted fork slots")]
+    MissingLastVotedForkSlots,
+    #[error("Hash parse error: {0}")]
+    ParseHashError(#[from] solana_sdk::hash::ParseHashError),
+    #[error("Pubkey parse error: {0}")]
+    PubkeyParseError(#[from] solana_sdk::pubkey::ParsePubkeyError),
+    #[error("Gossip push RestartLastVotedForkSlotsError: {0}")]
+    RestartLastVotedForkSlotsError(#[from] RestartLastVotedForkSlotsError),
+    #[error("Unexpected state: {0:?}")]
+    UnexpectedState(wen_restart_proto::State),
+}
+
+pub(crate) fn send_restart_last_voted_fork_slots(
     last_vote: VoteTransaction,
     blockstore: Arc<Blockstore>,
     cluster_info: Arc<ClusterInfo>,
-) -> Result<(), Box<dyn std::error::Error>> {
-    // repair and restart option does not work without last voted slot.
-    let last_vote_slot = last_vote
-        .last_voted_slot()
-        .expect("wen_restart doesn't work if local tower is wiped");
-    let mut last_vote_fork: Vec<u64> = AncestorIterator::new_inclusive(last_vote_slot, &blockstore)
-        .take(MAX_SLOTS_PER_ENTRY)
-        .collect();
+    progress: &mut WenRestartProgress,
+) -> Result<(), WenRestartError> {
+    let last_voted_fork_slots;
+    let last_vote_hash;
+    match &progress.my_last_voted_fork_slots {
+        Some(my_last_voted_fork_slots) => {
+            last_voted_fork_slots = my_last_voted_fork_slots.last_voted_fork_slots.clone();
+            last_vote_hash = Hash::from_str(&my_last_voted_fork_slots.last_vote_bankhash).unwrap();
+        }
+        None => {
+            // repair and restart option does not work without last voted slot.
+            if let VoteTransaction::Vote(ref vote) = last_vote {
```

Comment:
Will this error for validators who have a default tower at startup and haven't voted yet?

Reply:
The default tower has an empty slots list, so it will not fail on line 91, instead it will fail on line 88.

Reply:
Is that what we want or do we want validators with a default tower to be able to perform this restart as well?

Reply:
Hmm, if you have no last vote information, not sure you can contribute anything to the restart. I guess those guys can copy the generated snapshot like today. Or we can try to use information from the ledger if tower is not available?

Reply:
I think that's fine, we should have a clear error message telling those guys they need to wait then. 

I think it would be a better user experience if you verified the tower beforehand and printed a clear error message  if the tower is not valid/missing

Reply:
That's a good point, I added some error log there, see if this helps?